zx/coerce.go

```diff
@@ -51,53 +56,145 @@ func CoerceToFloat64(in zng.Value) (float64, bool) {
 	return out, true
 }
 
-// CoerceToInt attempts to convert a value to an integer.  Int, Count,
-// and Port can are all translated to an Int with the same native
-// value while a Double is converted only if the double is an integer.
+// CoerceToInt and CoerceToUint attempt to convert a value to the given
+// integer type.  Integer types (byte, int*, uint*) can be coerced
+// to another integer type if the value is in range.  float64 may
+// be coerced to an integer type only if the value is equivalent.
 // Time and Intervals are converted to an Int as their nanosecond
-// values. The resulting coerced value is written to out, and true is
-// returned. If the value cannot be coerced, then false is returned.
-func CoerceToInt(in zng.Value) (int64, bool) {
-	var out int64
+// values.
+func CoerceToInt(in zng.Value, typ zng.Type) (zng.Value, bool) {
+	var i int64
 	var err error
-	body := in.Bytes
-	switch in.Type.ID() {
+
+	switch in.Type.(type) {
```

Comment:
Doing this is going to break aliases.

Reply:
can you elaborate?

Reply:
What if you pass an aliased type into here?

Reply:
You have to either unwrap it or use `in.Type.ID()`

Reply:
> What if you pass an aliased type into here?

I beileve it is broken, whether we use `.(type)` or `.ID()`.  There's a separate PR that addresses this.

Reply:
In master, calls to `.ID()` on an alias will recurse to the original primitive type of the alias (see https://github.com/brimsec/zq/blob/master/zng/alias.go#L22). This does the same thing as the function you created in `AliasedType` in #326 .

Reply:
I had switched these coercion functions in the original alias pr to use id's instead of types for this very purpose.

Reply:
> I had switched these coercion functions in the original alias pr to use id's instead of types for this very purpose.

Great, it would have been nice to have handled compare.go as well and to have added tests to prevent future regressions.

Reply:
Yep, I thought I had handled all cases, but forgot about compare.go.

Reply:
FWIW I did add a test for groupby (https://github.com/brimsec/zq/pull/276/files#diff-9fcad515342ab12c2e5bf1a1d096caa5). Had I considered compare as well (which I didn't) I would have add a test there.